tokenserver/assignment/sqlnode/sql.py

```diff
@@ -459,16 +441,39 @@ def get_user_records(self, service, email):
     def get_old_user_records(self, service, grace_period=-1, limit=100,
                              offset=0):
         """Get user records that were replaced outside the grace period."""
+        nodes = self._get_nodes_table(service)
+        users = self._get_users_table(service)
+        service = self._get_service_id(service)
+
         if grace_period < 0:
             grace_period = 60 * 60 * 24 * 7  # one week, in seconds
         grace_period = int(grace_period * 1000)  # convert seconds -> millis
-        params = {
-            "service": service,
-            "timestamp": get_timestamp() - grace_period,
-            "limit": limit,
-            "offset": offset
-        }
-        res = self._safe_execute(_GET_OLD_USER_RECORDS_FOR_SERVICE, **params)
+        timestamp = get_timestamp() - grace_period
+
+        query = select([users.c.uid,
+                        users.c.email,
+                        users.c.generation,
+                        users.c.keys_changed_at,
+                        users.c.client_state,
+                        nodes.c.node,
+                        nodes.c.downed,
+                        users.c.created_at,
+                        users.c.replaced_at])
+        query = query.with_hint(users, 'USE INDEX (replaced_at_idx)',
+                                dialect_name='mysql')
+        query = query.select_from(users.outerjoin(
+            nodes,
+            users.c.nodeid == nodes.c.id))
```

Comment:
I am **so** totally pretending to have a clue what I'm doing here here ;-)

Reply:
Heh, yep, looks like sqlalchemy magic to me 👍